src/YHttpServer.cc

```diff
@@ -50,54 +51,117 @@ int YHttpServer::port_num()
     // allow only loop back connection ("http://localhost" only)
-    return htonl(INADDR_LOOPBACK);
+    yuiMilestone() << "Only localhost connections allowed." << std::endl;
+    return false;
+}
+
+// return the IPv4 listen address (localhost or all)
+in_addr_t listen_address_v4(bool remote)
+{
+    return (remote) ? INADDR_ANY : htonl(INADDR_LOOPBACK);
+}
+
+// return the IPv6 listen address (localhost or all)
+struct in6_addr listen_address_v6(bool remote)
+{
+    return (remote) ? in6addr_any : in6addr_loopback;
 }
 
-YHttpServer::YHttpServer() : server(nullptr), redraw(false)
+YHttpServer::YHttpServer() : server_v4(nullptr), server_v6(nullptr), redraw(false)
 {
     _yserver = this;
+
+    // authorization user set?
+    if (const char *user = getenv(YUI_AUTH_USER))
+    {
+        auth_user = user;
+        // unset to not leak to the child processes
+        unsetenv(YUI_AUTH_USER);
+    }
+
+    // authorization password set?
+    if (const char *pwd = getenv(YUI_AUTH_PASSWD))
+    {
+        auth_passwd = pwd;
+        // unset to not leak to the child processes
+        unsetenv(YUI_AUTH_PASSWD);
+    }
+
+    if (auth_user.empty() && auth_passwd.empty())
+        yuiWarning() << "User authentication not configured! " \
+            "You can pass it via environment variables " <<
+            YUI_AUTH_USER << " and " << YUI_AUTH_PASSWD <<std::endl;
 }
 
 YHttpServer::~YHttpServer()
 {
-    if (server) {
-        yuiMilestone() << "Stopping HTTP server" << std::endl;
-        MHD_stop_daemon(server);
+    yuiMilestone() << "Finishing the REST API HTTP server..." << std::endl;
+
+    if (server_v4) {
+        yuiMilestone() << "Stopping IPv4 HTTP server" << std::endl;
+        MHD_stop_daemon(server_v4);
+    }
+
+    if (server_v6) {
+        yuiMilestone() << "Stopping IPv6 HTTP server" << std::endl;
+        MHD_stop_daemon(server_v6);
     }
 }
 
 YHttpServerSockets YHttpServer::sockets()
 {
     fd_set rs, ws, es;
-    FD_ZERO (&rs);
-    FD_ZERO (&ws);
-    FD_ZERO (&es);
     int max = 0;
 
     YHttpServerSockets ret;
 
-    if (MHD_YES != MHD_get_fdset(server, &rs, &ws, &es, &max))
+    if (server_v4)
     {
-        yuiError() << "Cannot read the FD set!" << std::endl;
-        return ret;
+        FD_ZERO (&rs);
+        FD_ZERO (&ws);
+        FD_ZERO (&es);
+
+        if (MHD_YES != MHD_get_fdset(server_v4, &rs, &ws, &es, &max))
+        {
+            yuiError() << "Cannot read the IPv4 FD set!" << std::endl;
+            return ret;
+        }
+
+        for(int i = 0; i <= max; ++i) {
+            if (FD_ISSET(i, &rs)) ret.add_read(i);
+            if (FD_ISSET(i, &ws)) ret.add_write(i);
+            if (FD_ISSET(i, &es)) ret.add_exception(i);
+        }
     }
 
-    for(int i = 0; i <= max; ++i) {
-        if (FD_ISSET(i, &rs)) ret.add_read(i);
-        if (FD_ISSET(i, &ws)) ret.add_write(i);
-        if (FD_ISSET(i, &es)) ret.add_exception(i);
+    if (server_v6)
```

Comment:
There is quite some duplication here which will hit us in the face in the long run.

The real difference is the first parameter to `MHD_get_fdset()`, either `server_v4` or `server_v6`. Can both be active at the same time? Or none of them? If it's either one or the other, this could be simplified to

```C++
        FD_ZERO (&rs);
        FD_ZERO (&ws);
        FD_ZERO (&es);

        if (MHD_YES != MHD_get_fdset(server_v4 ? server_v4 : server_v6, &rs, &ws, &es, &max))
        {
            yuiError() << "Cannot read the FD set!" << std::endl;
            return ret;
        }

        for(int i = 0; i <= max; ++i) {
            if (FD_ISSET(i, &rs)) ret.add_read(i);
            if (FD_ISSET(i, &ws)) ret.add_write(i);
            if (FD_ISSET(i, &es)) ret.add_exception(i);
        }
    }

```

Reply:
>  Can both be active at the same time? Or none of them?

It depends on the network setup, if IPv6 is disabled then only the IPv4 part will be running. On the other hand in a pure IPv6 network the IPv4 server will not be active. (It will actually fail to start but we currently do not care, maybe we should fail if none of them succeed. See [that FIXME](https://github.com/libyui/libyui-rest-api/pull/2/files#diff-eabee18fc4d2ca5652cbbbd07604ee01R337).)